File: 28_jQuery/ch05/ajax_handy.go

package main

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"log"
	"os"
)

var bs = []byte(
	`[
		{
			"term": "BACCHUS",
			"part": "n.",
			"definition": "A convenient deity invented by the...",
			"quote": [
				"Is public worship, then, a sin,",
				"That for devotions paid to Bacchus",
				"The lictors dare to run us in,",
				"And resolutely thump and whack us?"
			],
			"author": "Jorace"
		},
		{
			"term": "BACKBITE",
			"part": "v.t.",
			"definition": "To speak of a man as you find him when..."
		}
	]`,
)

var cs = []byte(
	`[
		{
			"term": "BACCHUS",
			"part": "n.",
			"definition": "A convenient deity invented by the...",
			"quote": [
				"Is public worship, then, a sin,",
				"That for devotions paid to Bacchus",
				"The lictors dare to run us in,",
				"And resolutely thump and whack us?"
			],
			"author": "Jorace"
		},
		{
			"term": "BACKBITE",
			"part": "v.t.",
			"definition": "To speak of a man as you find him when..."
		},
		{
			"term": "BEARD",
			"part": "n.",
			"definition": "The hair that is commonly cut off by..."
		}
	]`,
)

func init() {
	log.SetFlags(log.LstdFlags|log.Lshortfile|log.Lmicroseconds)
}

func main() {
	router := gin.Default()

	pwd, _ := os.Getwd()

	hmp := pwd + "/html"
	csp := pwd + "/css"
	jsp := pwd + "/js"

	cssfile := "/ajax_handy.css"
	jsfile := "/ajax_handy.js"

	router.StaticFile(cssfile, csp + cssfile)
	router.StaticFile(jsfile, jsp + jsfile)

	xmlfile := "/ajax/get.xml"
	router.StaticFile(xmlfile, pwd + xmlfile)
	jsofile := "/ajax/getJSON.json"
	router.StaticFile(jsofile, pwd + jsofile)
	jstfile := "/ajax/getScript.js"
	router.StaticFile(jstfile, pwd + jstfile)
	hmtfile := "/ajax/load.html"
	router.StaticFile(hmtfile, pwd + hmtfile)

	router.LoadHTMLGlob(hmp + "/*")

    router.GET("/", func(c *gin.Context) {
        c.HTML(http.StatusOK, "ajax_handy.html", nil)
    })

    router.GET("/get", func(c *gin.Context) {
        c.String(http.StatusOK, string(bs))
	})

    router.POST("/post", func(c *gin.Context) {
		appId := c.PostForm("appId")
		userId := c.PostForm("userId")
		log.Printf("appId=%s,userId=%s", appId, userId)
		c.String(http.StatusOK, string(cs))
	})

	router.Run(":8080")
}

File: 28_jQuery/ch05/js/ajax_handy.js

// $.get(): 从服务器文件(.xml)中获取数据，于 B 端解析并嵌入到页面中
$(document).ready(function() {
    $('#get-xml .trigger').click(function(event) {
        event.preventDefault();
        $.get('ajax/get.xml', function(data) {
            $('#get-xml #dictionary').empty();
            $(data).find('entry').each(function() {
                var $entry = $(this);
                var html = '<div class="entry">';
                html += '<h3 class="term">' + $entry.attr('term');
                html += '</h3>';
                html += '<div class="part">' + $entry.attr('part');
                html += '</div>';
                html += '<div class="definition">';
                html += $entry.find('definition').text();
                var $quote = $entry.find('quote');
                if ($quote.length) {
                    html += '<div class="quote">';
                    $quote.find('line').each(function() {
                        html += '<div class="quote-line">';
                        html += $(this).text() + '</div>';
                    });
                    if ($quote.attr('author')) {
                        html += '<div class="quote-author">';
                        html += $quote.attr('author') + '</div>';
                    }
                    html += '</div>';
                }
                html += '</div>';
                html += '</div>';
                $('#get-xml #dictionary').append($(html));
            });
        });
    });
});

// $.getJSON(): 从服务器文件(.json)中获取数据，于 B 端解析并嵌入到页面中
$(document).ready(function() {
    $('#get-json .trigger').click(function(event) {
        event.preventDefault();
        $.getJSON('ajax/getJSON.json', function(data) {
            var html = '';
            $.each(data, function(entryIndex, entry) {
                html += '<div class="entry">';
                html += '<h3 class="term">' + entry.term + '</h3>';
                html += '<div class="part">' + entry.part + '</div>';
                html += '<div class="definition">';
                html += entry.definition;
                if (entry.quote) {
                    html += '<div class="quote">';
                    $.each(entry.quote, function(lineIndex, line) {
                        html += '<div class="quote-line">' + line + '</div>';
                    });
                    if (entry.author) {
                        html += '<div class="quote-author">' + entry.author + '</div>';
                    }
                    html += '</div>';
                }
                html += '</div>';
                html += '</div>';
            });
            $('#get-json #dictionary').html(html);
        });
    });
});

// $.getScript(): 从服务器文件(.js)中获取数据，于 B 端解析并嵌入到页面中
$(document).ready(function() {
    $('#get-script .trigger').click(function(event) {
        event.preventDefault();
        $.getScript('/ajax/getScript.js');
    });
});

// $.load(): 从服务器文件(.html)中获取数据，并嵌入到页面中
$(document).ready(function() {
    $('#load .trigger').click(function(event) {
        event.preventDefault();
        $('#load #dictionary').load('/ajax/load.html');
    });
});

// $.get(): 获取数据
$(document).ready(function(){
    $('#get .trigger').click(function(event) {
        $.ajax({
            type: "GET",
            url: "/get",
            dataType: "json",
            success: function(data) {
                jdata = JSON.stringify(data);
                var html = '';
                $.each(JSON.parse(jdata), function(entryIndex, entry) {
                    html += '<div class="entry">';
                    html += '<h3 class="term">' + entry.term + '</h3>';
                    html += '<div class="part">' + entry.part + '</div>';
                    html += '<div class="definition">';
                    html += entry.definition;
                    if (entry.quote) {
                        html += '<div class="quote">';
                        $.each(entry.quote, function(lineIndex, line) {
                            html += '<div class="quote-line">' + line + '</div>';
                        });
                        if (entry.author) {
                            html += '<div class="quote-author">' + entry.author + '</div>';
                        }
                        html += '</div>';
                    }
                    html += '</div>';
                    html += '</div>';
                });
                $('#get #dictionary').append(html);

                return false;
            }
        });
    });
});

// $.post(): 发送并获取数据
$(document).ready(function(){
    $('#post .trigger').click(function(event) {
        $.ajax({
            type: "POST",
            url: "/post",
            dataType: "json",
            data: {
                appId: "101",
                userId: "xiaoming",
            },
            contentType: "application/x-www-form-urlencoded",
            success: function(data) {
                jdata = JSON.stringify(data);
                var html = '';
                $.each(JSON.parse(jdata), function(entryIndex, entry) {
                    html += '<div class="entry">';
                    html += '<h3 class="term">' + entry.term + '</h3>';
                    html += '<div class="part">' + entry.part + '</div>';
                    html += '<div class="definition">';
                    html += entry.definition;
                    if (entry.quote) {
                        html += '<div class="quote">';
                        $.each(entry.quote, function(lineIndex, line) {
                            html += '<div class="quote-line">' + line + '</div>';
                        });
                        if (entry.author) {
                            html += '<div class="quote-author">' + entry.author + '</div>';
                        }
                        html += '</div>';
                    }
                    html += '</div>';
                    html += '</div>';
                });
                $('#post #dictionary').append(html);

                return false;
            }
        });
    });
});